### PR TITLE
kanidm: 1.1.0-alpha.11 -> 1.1.0-alpha.12

### PR DESCRIPTION
**Pull request by bjornfor**
###### Description of changes

The NixOS VM test requires more work -- kanidm has changed its CLI interface and
the test currently fails.

If anyone want to take over, please do.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/kanidm/default.nix [hunk 1]
```diff
@@ -17,24 +17,28 @@ let
...
     outputHashes = {
-      "tracing-forest-0.1.4" = "sha256-ofBLxSzZ5SYy8cbViVUa6VXKbOgd8lt7QUYhL0BW6I4=";
+      "tracing-forest-0.1.5" = "sha256-L6auSKB4DCnZBZpx7spiikhSOD6i1W3erc3zjn+26Ao=";
     };
   };
 
   KANIDM_BUILD_PROFILE = "release_nixos_${arch}";
 
+  patches = [
+    ./cpu-opt-level.patch
```
Comment (erictapen) [inline pkgs/servers/kanidm/default.nix:39]: Could you include this via fetchpatch (e.g. with `https://github.com/bjornfor/kanidm/commit/5d7ecc83d69bcc23da439abb8674a6d7da0951f6.patch`)? This is more common for patches that we try to bring upstream.

I can't really say wether this change is desirable at all, according to this [discussion](https://github.com/kanidm/kanidm/pull/1590#issuecomment-1529576062).

Comment (bjornfor) [reply pkgs/servers/kanidm/default.nix:39]: Maybe we should drop the patch and use CPU optimization `none` instead?

Comment (erictapen) [reply pkgs/servers/kanidm/default.nix:39]: Yeah, that sounds easier to me tbh. If anybody cares a lot about performance, they could always reintroduce a solution again.

**Comment by felixsinger** [comment]
> The NixOS VM test requires more work -- kanidm has changed its CLI interface and the test currently fails.

Could you please tell which tests or what exactly needs rework?

**Comment by bjornfor** [comment]
> Could you please tell which tests or what exactly needs rework?

I think it currently (tip og this PR) fails because "kanidm-unix status" fails.

**Comment by erictapen** [comment]
I think the remaining error is that there is that `kanidm-unix status` doesn't connect to the daemon correctly. When setting `RUST_LOG=trace`, one can see that the problem is some cryptic serde_json deserialization error, so it might be not so trivial.

```
vm-test-run-kanidm> (finished: waiting for unit kanidm-unixd.service, in 0.21 seconds)
vm-test-run-kanidm> client: must succeed: kanidm-unix status
vm-test-run-kanidm> (finished: must succeed: kanidm-unix status, in 0.18 seconds)
vm-test-run-kanidm> 2023-05-06T16:47:43.788897Z ERROR kanidm_unix: Error: unexpected response -> Error
vm-test-run-kanidm> client: must succeed: kanidm-unix status | grep working!
vm-test-run-kanidm> client: output:
vm-test-run-kanidm> cleanup
vm-test-run-kanidm> kill machine (pid 6)
vm-test-run-kanidm> client # qemu-kvm: terminating on signal 15 from pid 4 (/nix/store/6jx4nc20v3cfphlbypqid30q41i5vh5x-python3-3.10.11/bin/python3.10)
vm-test-run-kanidm> kill machine (pid 17)
vm-test-run-kanidm> server # qemu-kvm: terminating on signal 15 from pid 4 (/nix/store/6jx4nc20v3cfphlbypqid30q41i5vh5x-python3-3.10.11/bin/python3.10)
vm-test-run-kanidm> (finished: cleanup, in 0.18 seconds)
vm-test-run-kanidm> Traceback (most recent call last):
vm-test-run-kanidm>   File "/nix/store/iadsrhgccynfpn96qsn9yq9hrplckhjp-nixos-test-driver-1.1/bin/.nixos-test-driver-wrapped", line 9, in <module>
vm-test-run-kanidm>     sys.exit(main())
vm-test-run-kanidm>   File "/nix/store/iadsrhgccynfpn96qsn9yq9hrplckhjp-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/__init__.py", line 112, in main
vm-test-run-kanidm>     driver.run_tests()
vm-test-run-kanidm>   File "/nix/store/iadsrhgccynfpn96qsn9yq9hrplckhjp-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/driver.py", line 142, in run_tests
vm-test-run-kanidm>     self.test_script()
vm-test-run-kanidm>   File "/nix/store/iadsrhgccynfpn96qsn9yq9hrplckhjp-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/driver.py", line 138, in test_script
vm-test-run-kanidm>     exec(self.tests, symbols, None)
vm-test-run-kanidm>   File "<string>", line 20, in <module>
vm-test-run-kanidm>   File "/nix/store/iadsrhgccynfpn96qsn9yq9hrplckhjp-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/machine.py", line 598, in succeed
vm-test-run-kanidm>     raise Exception(f"command `{command}` failed (exit code {status})")
vm-test-run-kanidm> Exception: command `kanidm-unix status | grep working!` failed (exit code 1)
vm-test-run-kanidm> kill vlan (pid 5)
```

**Comment by RaitoBezarius** [comment]
I pushed a fix because @mweinelt wanted one, but I'm not yet able to take over this package or this PR. Please feel free to finish up the work and take it to upstream regarding the readiness problem.

**Comment by mweinelt** [comment]
Created an issue about the socket being present, but erroring out. https://github.com/kanidm/kanidm/issues/1677

File: pkgs/servers/kanidm/default.nix (hunk 1 above)
Comment (erictapen) [reply pkgs/servers/kanidm/default.nix:39]: This is not resolved yet. Either you'd need to `fetchpatch` https://github.com/kanidm/kanidm/commit/59c6723f7dfb2266eae45c3b2ddd377872a7a113 or change our build profile to  `none`.

**Comment by erictapen** [comment]
I think as https://github.com/kanidm/kanidm/issues/1677 is resolved now, we could just try to fetchpatch https://github.com/kanidm/kanidm/commit/466acb472934e625875925efc0f00a46820e8bf6 and revert @RaitoBezarius changes to the test logic.

**Comment by mweinelt** [comment]
The patch doesn't apply cleanly, and I don't think it's worth vendoring it. I will clean up this PR in a bit.

**Comment by mweinelt** [comment]
- Replaced with vendored patch with a fetchpatch of the patch that went upstream
- Moved all tests into subtests
- Squashed Ryan's fix into the test update change
- Mention co-authors everywhere
- Rebased onto master

File: pkgs/servers/kanidm/default.nix (hunk 1 above)
Comment (mweinelt) [reply pkgs/servers/kanidm/default.nix]: Went fetchpatching the referenced commit, given that this is what went upstream.

**Comment by bjornfor** [comment]
Commits must be squashed before landing.

**Comment by mweinelt** [comment]
You mean we should squash the nixos test changes into the package version bump? I don't think we have to do that.

**Comment by mweinelt** [comment]
The test is failing on ofborg for aarch64-linux, but succeeds for me on the aarch64 community builder, which should be the same machine.

**Comment by bjornfor** [comment]
> You mean we should squash the nixos test changes into the package version bump? I don't think we have to do that.

Yes, that's what I meant. I think if we don't squash then we're leaving the repo in a broken state after the first commit.

Perhaps in this instance it doesn't matter a lot, but over time it gets annoying if we allow PRs to introduce broken in-between states -- it hurts git bisect hunting for real bugs.

**Comment by mweinelt** [comment]
Fine with me. Squashed.

**Review by erictapen — APPROVED** [review]
My desktop PC with 12GB RAM is going OOM wile building kanidm :/
So I can't verify this and I find the errors reported on ofborg weird. But as @mweinelt reports succesful tests, and the code looks good, I'd say this is ready to be merged.

**Comment by mweinelt** [comment]
Tests succeed on x86-64-linux and aarch64-linux for me.

**Pull request merged**